docs/content/basics/testing_theory.md

```diff
@@ -1,3 +1,57 @@
+<br/>It is also worth noting that there will be an accent on UI tests, because we are in CookBook for UI tests.
+
+# Where it all starts
+<i>Why do I need testing?</i> Users of your application/system/whatever should receive a quality product. 
+<br/>Thus, testing is about ensuring the quality of the product. Testing also includes creating a test plan, creating/conducting tests themselves and analyzing results of testing.
+
+Let's focus on the topic of creating tests. To begin with, let's see the testing pyramid.
+
+<image here>
+
+The width of each block is actually the ratio of the number of different types of tests to each other. 
+For example, a lot of Unit tests is usually considered correct in the pyramid, but much less UI tests. 
+<br/><br/><b>This is due to two key parameters - stability and the cost of supporting each type of testing.</b>
+<br/>Unit tests have the highest stability, they are the fastest and they have the lowest cost of support, however, using Unit tests you won't be able to check the login flow of your application (which for example contains 5 screens).
+<br/><br/>We will reveal each type of testing in more detail, for this we will go from the bottom up.
+
+!!! Disclaimer
+
+    <b>White-box testing</b> is a type of testing in which we have full access to the implementation and can interact with it. We know which output data will be with given input data.
+    <br/><b>Black-box testing</b> is a type of testing in which we don't have access to the implementation and cannot interact with it, however, we know which output data should be with given input data.
+    <br/><b>Gray-box testing</b> is a rare type of testing when you have partial access to the implementation (for example, not to all entities that being tested). At the same time, we know what output data will be with given input data.
+
+## Unit tests
+These are tests that check the operation of specific entities (for example, a repository class or a ViewModel) and are directly related to the implementation, which gives us the opportunity to test on the principle of a white box.
+<br/>They have high stability, because in order to break them, it is necessary to change the specific entity that they are testing.
+
+## Integration tests
+This type of tests is already a little less stable (than Unit tests), since we are going to test the interaction of 2 (or more) entities at once at the same time, but it still allows you to write in a White-box style. Now, having changed only one entity out of 2, there is a risk of failing integration tests.
+<br/>It is by no means impossible to ignore these tests, since the Unit tests passed give us an understanding of what specific entities work, but how these entities work together we can only find out inside the application or using Integration testing.
+
+!!! Info
+    
+    Examples of Integration tests in Android include database testing.
+
+## UI tests
```

Comment:
I'd also mention that UI tests they are also integration tests at the same time

Reply:
Actually in the android world I understand the separation better like "Instrumentation Tests" (running on an emulator) vs. "Local tests" (running on the JVM)
Alex Zukhovich has a great article about it:
https://alexzh.com/local-and-instrumentation-tests-in-android/

Then you can clarify Instrumentation tests into several groups:
e.g regarding Focus:
1. UI tests -> Focus on testing the WHAT (interaction/navigation): WHAT is displayed when interacting with the screen elements
2. Snapshot tests -> Focus on testing the HOW (visuals): HOW a view is displayed under a given state.
3. Non-UI tests -> Focus on testing classes strongly dependent on the Android  Framework (e.g. context, shared preferences, etc.)

Regarding granularity, UI tests  tests can be
1. E2E Tests: No mocks at all. Test the system as a whole. They test full user stories. They are the most unstable since there is no way to guarantee stability on some system parts like internet connection.
2. Integration tests: They test part of user stories, and span no more than a couple of screens i.e Activities or Fragments. They might also have other parts of the Android system mocked (e.g. network layer for better stability and speed) 
3. Unit tests: They only test a View/Activity/Fragment in isolation

Regarding Snapshot tests can be:
1. Full Screen Tests: They test all the views in a screen as a whole. They might also have other parts of the Android system mocked (e.g. network layer for better stability and speed).
2. View Unit Tests: They test how a small View unit displays under a given state. A small view unit could be a Dialog, a ViewHolder, a Composable, custom View... basically every Ui component we can drive to a given state without too much hassle (e.g. no data fetching) 

Therefore we can have:
 1. Ui E2E Test,
 2. Ui IntegrationTest
 3. Ui Unit Test
 
 and
 1. Full Screen Snapshot test
 2. View Unit Snapshot test
-----
On the other hand we have the Local Tests. Again they can be classified differently. For example, regarding focus:
1. Ui tests: They run using Robolectric, which mocks most of the Android System to let them run on the JVM. It comes with some limitations (e.g. Robolectric does not support Activity/fragment jumping).
2. Snapshot tests: They run using Paparazzi, which uses the same resources as Layout Validation in Android Studio to display a View without an Android device or emulator
3. Non-UI tests: They tests functions without dependencies on the Android frameworks

Regarding granularity, local Ui Tests can only be unit tests, since Robolectric does not support activity/fragment jumping.
1. Local Ui Unit tests

Snapshot tests have the same granularity when running as instrumentation tests as when doing it locally
1. Local Full Screen Snapshot Test
2. Local View Unit Snapshot Test

Non-UI tests can be either
1. Local Integration tests: Test how several functions work together. E.g. calling different methods in a ViewModel call the expected methods in the View interface
2. Local Unit tests: They test a single function

-----
Apart Fromm  the White/black/grey box testing (which is another way to classify tests by access to implementation), this kind of structure would be my proposal, maybe accompanied by a graph, and adding some Pros and Cons for each approach?
What are your thoughts?

Reply:
@sergio-sastre looks great, thank you! Really liked it

About separation, I think that somehow we can replace `local test` to something like:

`Unit test:` Can run on JVM only, abstracted from an `Android SDK`
`Instrumented test:` They have a knowledge about `Android SDK`. 
Because of that, we can run them only on Emulators or Physical Devices.


@v1sar Looks like we can combine everything together

Reply:
discussed in private that @sergio-sastre will add this info in separate PR.